File: src/de/ax/powermode/power/element/PowerSpark.scala
/*
 * Copyright 2015 Baptiste Mesta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.ax.powermode.power.element

import java.awt._

import de.ax.powermode.power.ElementOfPower
import de.ax.powermode.{PowerColor, Util}

/**
  * @author Baptiste Mesta
  */
case class PowerSpark(var x: Float, var y: Float, dx: Float, var dy: Float, size: Float, val initLife: Long, color: PowerColor, gravityFactor: Float) extends ElementOfPower {
  val life = System.currentTimeMillis() + initLife

  def update(delta: Float): Boolean = {
    dy += (0.07f * gravityFactor) * delta
    x += dx * delta
    y += dy * delta
    !alive
  }

  def render( g: Graphics, dxx: Int, dyy: Int) {
    if (alive) {
      val g2d: Graphics2D = g.create.asInstanceOf[Graphics2D]
      g2d.setColor(new Color(color._1, color._2, color._3, Util.alpha(color._4)))
      g2d.fillOval((dxx + x - (size / 2)).toInt, (dyy + y - (size / 2)).toInt, size.toInt, size.toInt)
      g2d.dispose()
    }
  }

}

File: src/com/nmeylan/powermode/power/element/PowerSpark.java
package com.nmeylan.powermode.power.element;

import com.nmeylan.powermode.Util;
import com.nmeylan.powermode.power.ElementOfPower;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;

public class PowerSpark implements ElementOfPower {

  private float x;
  private float y;
  private float dx;
  private float dy;
  private int size;
  private long initLife;
  private long life;
  private float[] colors;
  private float gravityFactor;

  public PowerSpark(float x, float y, float dx, float dy, int size, long initLife, float[] colors, float gravityFactor) {
    this.x = x;
    this.y = y;
    this.dx = dx;
    this.dy = dy;
    this.size = size;
    this.initLife = initLife;
    this.colors = colors;
    this.gravityFactor = gravityFactor;
    this.life = System.currentTimeMillis() + initLife;
  }

  @Override
  public boolean update(double delta) {
    dy += (0.07f * gravityFactor) * delta;
    x += dx * delta;
    y += dy * delta;
    return !alive();
  }

  @Override
  public void render(Graphics g, int dxx, int dyy) {
    if (alive()) {
      Graphics2D g2d = (Graphics2D) g.create();
      g2d.setColor(new Color(colors[0], colors[1], colors[2], Util.alpha(colors[3])));
      g2d.fillOval((int)(dxx + x - (size / 2)), (int)(dyy + y - (size / 2)), size,size);
      g2d.dispose();
    }
  }

  @Override
  public long life() {
    return life;
  }

  @Override
  public long initLife() {
    return this.initLife;
  }
}
